concordium-consensus/src/Concordium/KonsensusV1/LeaderElection.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE DataKinds #-}
+
+module Concordium.KonsensusV1.LeaderElection (
+    -- * Leader election
+    getLeader,
+
+    -- * Seed state
+    updateSeedStateForBlock,
+    updateSeedStateForEpoch,
+) where
+
+import Data.Serialize
+
+import qualified Concordium.Crypto.SHA256 as Hash
+import Concordium.Types
+import Concordium.Types.HashableTo
+import Concordium.Types.SeedState
+
+import Concordium.Crypto.VRF (proofToHash)
+import Concordium.GlobalState.BakerInfo
+import Concordium.KonsensusV1.Types
+
+-- |Compute the leader for a given round, given the set of bakers and leadership election nonce
+-- for the epoch. The leader is computed as follows:
+--
+--  * Compute the SHA256 hash of the nonce and round number.
+--  * Treating this as an big-endian unsigned integer, compute the remainder when divided by the
```

Comment:
Taking the remainder modulo total stake is different from the description in the bluepaper. Why is it done like this?

Reply:
The bluepaper deals with real numbers in the interval `[0,1)`. For the implementation we instead use integers in `[0 .. totalStake)`, where `totalStake` is the sum of the effective stakes of the bakers (accounting for (non-passive) delegation and stake limiting). We then use the remainder to map the hash into this interval. This is easier to work with, but should achieve the same effect.

Reply:
The real numbers would correspond to dividing the hash by 2^256. If you don't want to divide by a large number, the stakes could by multiplied by that. This would not have this bias discussed below. Would that be less efficient?

Reply:
If we divide by 2^256 and work with real arithmetic, it doesn't change the bias. To conceptualise it, let's shrink the numbers and suppose that the hash generates a value in the range `[0 .. 7]` and there are three bakers, each with stake 1.

If we consider as implemented, the bakers win on the inputs as follows:
1. `{x ∈ [0 .. 7] | x mod 3 = 0} = {0,3,6}`
2. `{x ∈ [0 .. 7] | x mod 3 = 1} = {1,4,7}`
3. `{x ∈ [0 .. 7] | x mod 3 = 2} = {2,5}`

If we consider dividing by 8 and then splitting in regions we get:
1. `{x ∈[0 .. 7] | 0 ≤ x < 1/3} = {0,1,2}`
2. `{x ∈[0 .. 7] | 1/3 ≤ x < 2/3} = {3,4,5}`
3. `{x ∈[0 .. 7] | 1/3 ≤ x < 1} = {6,7}`

Even though the mapping is different, the bias is the same.

Reply:
Thanks, makes sense. And the modulo thing is more efficient I assume?

We should probably change the bluepaper then and add some of your reasoning about the bias.